File: ForgeSwift/ForgeCore/Sources/ForgeCore/Intelligence/ScheduleCorrector.swift
import Foundation

/// An explicit wake target and the first morning it should hold.
///
/// Sleep-medicine guidance is a 15–20 minute shift per night — jet-lag style —
/// not a single jump. `CircadianRhythm.phase()` is the current body-clock;
/// this is the destination and the ratchet toward it.
public struct ScheduleGoal: Codable, Equatable, Sendable {
    public var targetWakeHour: Double
    public var cutoverStart: Date
    public var maxShiftMinutesPerNight: Double
    public var createdAt: Date
    public var isActive: Bool

    public static let defaultMaxShiftMinutes = 15.0
    public static let hardMaxShiftMinutes = 20.0

    public init(
        targetWakeHour: Double,
        cutoverStart: Date,
        maxShiftMinutesPerNight: Double = ScheduleGoal.defaultMaxShiftMinutes,
        createdAt: Date = Date(),
        isActive: Bool = true
    ) {
        self.targetWakeHour = CircadianRhythm.normalizedHour(targetWakeHour)
        self.cutoverStart = cutoverStart
        self.maxShiftMinutesPerNight = min(
            ScheduleGoal.hardMaxShiftMinutes,
            max(5, maxShiftMinutesPerNight)
        )
        self.createdAt = createdAt
        self.isActive = isActive
    }
}

/// Tonight's step on the ladder toward `ScheduleGoal`.
public struct ScheduleCorrectionStep: Equatable, Sendable {
    public var recommendedWakeHour: Double
    public var recommendedOnsetHour: Double
    public var shiftMinutesTonight: Double
    public var remainingGapMinutes: Double
    public var nightsRemainingEstimate: Int
    public var reachedTarget: Bool
    public var currentWakeHour: Double
    public var confidence: Double
    public var targetWakeHour: Double

    public init(
        recommendedWakeHour: Double,
        recommendedOnsetHour: Double,
        shiftMinutesTonight: Double,
        remainingGapMinutes: Double,
        nightsRemainingEstimate: Int,
        reachedTarget: Bool,
        currentWakeHour: Double,
        confidence: Double,
        targetWakeHour: Double
    ) {
        self.recommendedWakeHour = recommendedWakeHour
        self.recommendedOnsetHour = recommendedOnsetHour
        self.shiftMinutesTonight = shiftMinutesTonight
        self.remainingGapMinutes = remainingGapMinutes
        self.nightsRemainingEstimate = nightsRemainingEstimate
        self.reachedTarget = reachedTarget
        self.currentWakeHour = currentWakeHour
        self.confidence = confidence
        self.targetWakeHour = targetWakeHour
    }

    public var guidanceLine: String {
        let target = ScheduleCorrector.clockLabel(targetWakeHour)
        if reachedTarget {
            return "Wake target is \(target) — you're on it."
        }
        let minutes = Int(abs(shiftMinutesTonight).rounded())
        let direction = shiftMinutesTonight < 0 ? "earlier" : "later"
        let nights = max(1, nightsRemainingEstimate)
        return "Shifting \(minutes) min \(direction) toward \(target) — about \(nights) night\(nights == 1 ? "" : "s")."
    }

    public var coachingReply: String {
        let target = ScheduleCorrector.clockLabel(targetWakeHour)
        let bed = ScheduleCorrector.clockLabel(recommendedOnsetHour)
        if reachedTarget {
            return "You're already up at \(target). I'll keep bedtime near \(bed) so sleep holds tonight."
        }
        let minutes = Int(abs(shiftMinutesTonight).rounded())
        let direction = shiftMinutesTonight < 0 ? "earlier" : "later"
        return "I'll get you up at \(target) — \(minutes) min \(direction) tonight so sleep lands near \(bed)."
    }
}

/// Goal-directed phase advance/delay. Stateless: current phase in, tonight's
/// onset/wake out. Persistence lives in `ScheduleGoalStore`.
public enum ScheduleCorrector: Sendable {

    public static let reachedThresholdMinutes = 5.0

    /// Shortest signed hour delta on the clock, −12...12.
    public static func signedHourDelta(from: Double, to: Double) -> Double {
        var delta = CircadianRhythm.normalizedHour(to) - CircadianRhythm.normalizedHour(from)
        if delta > 12 { delta -= 24 }
        if delta < -12 { delta += 24 }
        return delta
    }

    public static func clockLabel(_ hour: Double) -> String {
        let normalized = CircadianRhythm.normalizedHour(hour)
        var whole = Int(normalized)
        var minute = Int(((normalized - Double(whole)) * 60).rounded())
        if minute >= 60 {
            minute = 0
            whole = (whole + 1) % 24
        }
        let suffix = whole >= 12 ? "pm" : "am"
        let h12 = whole % 12 == 0 ? 12 : whole % 12
        return String(format: "%d:%02d %@", h12, minute, suffix)
    }

    public static func tonight(
        goal: ScheduleGoal,
        nights: [CircadianRhythm.Night],
        now: Date = Date(),
        calendar: Calendar = .current
    ) -> ScheduleCorrectionStep? {
        guard goal.isActive else { return nil }
        guard let phase = CircadianRhythm.phase(from: nights, calendar: calendar) else { return nil }

        let need = CircadianRhythm.sleepNeedHours(from: nights)
        let gapHours = signedHourDelta(from: phase.wakeHour, to: goal.targetWakeHour)
        let gapMinutes = gapHours * 60
        let reached = abs(gapMinutes) < reachedThresholdMinutes
        let maxShiftHours = goal.maxShiftMinutesPerNight / 60.0

        let startToday = calendar.startOfDay(for: now)
        let cutoverDay = calendar.startOfDay(for: goal.cutoverStart)
        let daysUntil = calendar.dateComponents([.day], from: startToday, to: cutoverDay).day ?? 0

        let shiftHours: Double
        if reached {
            shiftHours = 0
        } else if daysUntil > 0 {
            let planned = gapHours / Double(daysUntil)
            shiftHours = min(maxShiftHours, max(-maxShiftHours, planned))
        } else {
            shiftHours = min(maxShiftHours, max(-maxShiftHours, gapHours))
        }

        let tonightWake = CircadianRhythm.normalizedHour(phase.wakeHour + shiftHours)
        let tonightOnset = CircadianRhythm.normalizedHour(tonightWake - need)
        let remaining = abs(gapMinutes - shiftHours * 60)
        let nightsLeft: Int
        if reached {
            nightsLeft = 0
        } else {
            nightsLeft = max(1, Int(ceil(abs(gapHours) / max(0.01, maxShiftHours))))
        }

        return ScheduleCorrectionStep(
            recommendedWakeHour: tonightWake,
            recommendedOnsetHour: tonightOnset,
            shiftMinutesTonight: shiftHours * 60,
            remainingGapMinutes: remaining,
            nightsRemainingEstimate: nightsLeft,
            reachedTarget: reached,
            currentWakeHour: phase.wakeHour,
            confidence: phase.confidence,
            targetWakeHour: goal.targetWakeHour
        )
    }
}

public enum ScheduleGoalStore: Sendable {
    public static let defaultsKey = "forge.sleep.scheduleGoal.v1"

    public static func save(_ goal: ScheduleGoal, defaults: UserDefaults = .standard) {
        if let data = try? JSONEncoder().encode(goal) {
            defaults.set(data, forKey: defaultsKey)
        }
    }

    public static func load(defaults: UserDefaults = .standard) -> ScheduleGoal? {
        guard let data = defaults.data(forKey: defaultsKey),
              let goal = try? JSONDecoder().decode(ScheduleGoal.self, from: data),
              goal.isActive else { return nil }
        return goal
    }

    public static func clear(defaults: UserDefaults = .standard) {
        defaults.removeObject(forKey: defaultsKey)
    }
}

/// Conversational "up at 6am starting Monday" → `ScheduleGoal`.
public enum ScheduleGoalParser: Sendable {

    public static func isScheduleAsk(_ text: String) -> Bool {
        parse(text) != nil
    }

    public static func parse(
        _ text: String,
        now: Date = Date(),
        calendar: Calendar = .current
    ) -> ScheduleGoal? {
        let lower = text.lowercased()
        guard hasScheduleCue(lower) else { return nil }
        guard let hour = parseWakeHour(in: lower) else { return nil }
        let cutover = parseCutover(in: lower, now: now, calendar: calendar)
        return ScheduleGoal(
            targetWakeHour: hour,
            cutoverStart: cutover,
            createdAt: now
        )
    }

    private static func hasScheduleCue(_ lower: String) -> Bool {
        let cues = [
            "up at", "wake at", "wake me", "get up at", "be up at",
            "getting up at", "alarm at", "wake time", "start waking",
            "waking up at", "need to be up",
        ]
        return cues.contains { lower.contains($0) }
    }

    static func parseWakeHour(in lower: String) -> Double? {
        let pattern = #"(?:up at|wake(?: me)?(?: up)? at|get up at|be up at|getting up at|waking up at|alarm at|wake time(?: of)?|need to be up(?: at)?)\s+(\d{1,2})(?::(\d{2}))?\s*(a\.?m\.?|p\.?m\.?)?"#
        guard let regex = try? NSRegularExpression(pattern: pattern, options: []) else { return nil }
        let range = NSRange(lower.startIndex..<lower.endIndex, in: lower)
        guard let match = regex.firstMatch(in: lower, options: [], range: range),
              match.numberOfRanges >= 2,
              let hourRange = Range(match.range(at: 1), in: lower),
              var hour = Int(lower[hourRange]) else { return nil }
        var minute = 0
        if match.numberOfRanges >= 3, match.range(at: 2).location != NSNotFound,
           let minRange = Range(match.range(at: 2), in: lower) {
            minute = Int(lower[minRange]) ?? 0
        }
        var isPM: Bool?
        if match.numberOfRanges >= 4, match.range(at: 3).location != NSNotFound,
           let ampmRange = Range(match.range(at: 3), in: lower) {
            let stamp = lower[ampmRange]
            if stamp.contains("p") { isPM = true }
            if stamp.contains("a") { isPM = false }
        }
        hour = min(23, max(0, hour))
        minute = min(59, max(0, minute))
        if let isPM {
            if isPM, hour < 12 { hour += 12 }
            if isPM == false, hour == 12 { hour = 0 }
        } else if hour == 0 || hour == 24 {
            hour = 0
        } else if hour > 24 {
            return nil
        }
        // Bare "6" on a wake ask is 6am, not 18:00.
        if isPM == nil, hour >= 13, hour <= 23 {
            // already 24h
        } else if isPM == nil, hour == 12 {
            hour = 12
        }
        return Double(hour) + Double(minute) / 60.0
    }

    static func parseCutover(in lower: String, now: Date, calendar: Calendar) -> Date {
        let weekdays: [(Int, [String])] = [
            (1, ["sunday"]),
            (2, ["monday"]),
            (3, ["tuesday"]),
            (4, ["wednesday"]),
            (5, ["thursday"]),
            (6, ["friday"]),
            (7, ["saturday"]),
        ]
        for (code, names) in weekdays {
            if names.contains(where: { lower.contains($0) }) {
                return nextWeekday(code, from: now, calendar: calendar)
            }
        }
        return now
    }

    static func nextWeekday(_ weekday: Int, from now: Date, calendar: Calendar) -> Date {
        let start = calendar.startOfDay(for: now)
        for offset in 0..<8 {
            guard let day = calendar.date(byAdding: .day, value: offset, to: start) else { continue }
            if calendar.component(.weekday, from: day) == weekday { return day }
        }
        return start
    }
}

File: backend/infra/lambda/aria_core/schedule_corrector.py
"""ScheduleCorrector + ScheduleGoalParser — goal-directed wake-time phase
advance/delay, and conversational parsing of "up at 6am starting Monday"
into a schedule goal.

Python port of ForgeCore's ``ScheduleCorrector.swift``
(``ForgeSwift/ForgeCore/Sources/ForgeCore/Intelligence/ScheduleCorrector.swift``),
kept numerically identical to it. Every constant and formula below cites the
exact Swift source line range it mirrors. Depends on
``aria_core.circadian_rhythm`` for phase/sleep-need estimation and hour
normalization -- the same dependency the Swift source has on
``CircadianRhythm.swift``.

Sleep-medicine guidance is a 15-20 minute shift per night (jet-lag style),
not a single jump: ``ScheduleCorrector.tonight()`` takes the current phase
(from ``circadian_rhythm.phase()``) and a target, and ratchets toward it at
a capped nightly rate, spread evenly across the nights remaining until a
cutover date when one is set.

Ported: ``ScheduleGoal``, ``ScheduleCorrectionStep`` (including its
``guidance_line``/``coaching_reply`` copy, word for word), ``tonight()``,
``signed_hour_delta()``, ``clock_label()``, and the full
``ScheduleGoalParser`` (cue detection, the wake-hour regex, weekday cutover
parsing).

Not ported: ``ScheduleGoalStore`` (UserDefaults persistence, iOS-only --
same reasoning as ``QualityOfLifeLivingStore``/``SleepDepthBaselineStore``
in the earlier ports).

Stdlib only (aside from the sibling ``aria_core.circadian_rhythm`` import).
Deterministic (given `now`). Pure -- no I/O.
"""

from __future__ import annotations

import math
import re
from dataclasses import dataclass, field
from datetime import datetime, timedelta

from . import circadian_rhythm as cr


def _swift_round(value: float) -> int:
    """Swift's ``Double.rounded()`` default rule: round half away from
    zero, not Python's banker's-rounding ``round()``."""
    return math.floor(value + 0.5) if value >= 0 else math.ceil(value - 0.5)


# --- Goal ----------------------------------------------------------------
# Mirrors ScheduleGoal (Swift L8-34).

DEFAULT_MAX_SHIFT_MINUTES = 15.0
HARD_MAX_SHIFT_MINUTES = 20.0


@dataclass
class ScheduleGoal:
    """An explicit wake target and the first morning it should hold.

    Sleep-medicine guidance is a 15-20 minute shift per night -- jet-lag
    style -- not a single jump. `circadian_rhythm.phase()` is the current
    body-clock; this is the destination and the ratchet toward it."""

    target_wake_hour: float
    cutover_start: datetime
    max_shift_minutes_per_night: float = DEFAULT_MAX_SHIFT_MINUTES
    created_at: datetime = field(default_factory=datetime.now)
    is_active: bool = True

    def __post_init__(self) -> None:
        self.target_wake_hour = cr.normalized_hour(self.target_wake_hour)
        self.max_shift_minutes_per_night = min(
            HARD_MAX_SHIFT_MINUTES, max(5.0, self.max_shift_minutes_per_night)
        )


# --- Step ------------------------------------------------------------------
# Mirrors ScheduleCorrectionStep (Swift L37-91), including its guidance/
# coaching copy word for word.

@dataclass(frozen=True)
class ScheduleCorrectionStep:
    """Tonight's step on the ladder toward `ScheduleGoal`."""

    recommended_wake_hour: float
    recommended_onset_hour: float
    shift_minutes_tonight: float
    remaining_gap_minutes: float
    nights_remaining_estimate: int
    reached_target: bool
    current_wake_hour: float
    confidence: float
    target_wake_hour: float

    @property
    def guidance_line(self) -> str:
        target = clock_label(self.target_wake_hour)
        if self.reached_target:
            return f"Wake target is {target} — you're on it."
        minutes = _swift_round(abs(self.shift_minutes_tonight))
        direction = "earlier" if self.shift_minutes_tonight < 0 else "later"
        nights = max(1, self.nights_remaining_estimate)
        plural = "" if nights == 1 else "s"
        return f"Shifting {minutes} min {direction} toward {target} — about {nights} night{plural}."

    @property
    def coaching_reply(self) -> str:
        target = clock_label(self.target_wake_hour)
        bed = clock_label(self.recommended_onset_hour)
        if self.reached_target:
            return f"You're already up at {target}. I'll keep bedtime near {bed} so sleep holds tonight."
        minutes = _swift_round(abs(self.shift_minutes_tonight))
        direction = "earlier" if self.shift_minutes_tonight < 0 else "later"
        return f"I'll get you up at {target} — {minutes} min {direction} tonight so sleep lands near {bed}."


# --- Corrector -----------------------------------------------------------
# Mirrors ScheduleCorrector (Swift L95-171): stateless, current phase in,
# tonight's onset/wake out. Persistence lives in ScheduleGoalStore (not
# ported).

REACHED_THRESHOLD_MINUTES = 5.0


def signed_hour_delta(from_: float, to: float) -> float:
    """Shortest signed hour delta on the clock, -12...12."""
    delta = cr.normalized_hour(to) - cr.normalized_hour(from_)
    if delta > 12:
        delta -= 24
    if delta < -12:
        delta += 24
    return delta


def clock_label(hour: float) -> str:
    normalized = cr.normalized_hour(hour)
    whole = int(normalized)
    minute = _swift_round((normalized - whole) * 60)
    if minute >= 60:
        minute = 0
        whole = (whole + 1) % 24
    suffix = "pm" if whole >= 12 else "am"
    h12 = 12 if whole % 12 == 0 else whole % 12
    return f"{h12}:{minute:02d} {suffix}"


def tonight(goal: ScheduleGoal, nights: list[cr.Night],
            now: datetime | None = None) -> ScheduleCorrectionStep | None:
    if not goal.is_active:
        return None
    if now is None:
        now = datetime.now()
    phase = cr.phase(nights)
    if phase is None:
        return None

    need = cr.sleep_need_hours(nights)
    gap_hours = signed_hour_delta(phase.wake_hour, goal.target_wake_hour)
    gap_minutes = gap_hours * 60
    reached = abs(gap_minutes) < REACHED_THRESHOLD_MINUTES
    max_shift_hours = goal.max_shift_minutes_per_night / 60.0

    start_today = now.replace(hour=0, minute=0, second=0, microsecond=0)
    cutover_day = goal.cutover_start.replace(hour=0, minute=0, second=0, microsecond=0)
    days_until = (cutover_day - start_today).days

    if reached:
        shift_hours = 0.0
    elif days_until > 0:
        planned = gap_hours / days_until
        shift_hours = min(max_shift_hours, max(-max_shift_hours, planned))
    else:
        shift_hours = min(max_shift_hours, max(-max_shift_hours, gap_hours))

    tonight_wake = cr.normalized_hour(phase.wake_hour + shift_hours)
    tonight_onset = cr.normalized_hour(tonight_wake - need)
    remaining = abs(gap_minutes - shift_hours * 60)
    if reached:
        nights_left = 0
    else:
        nights_left = max(1, math.ceil(abs(gap_hours) / max(0.01, max_shift_hours)))

    return ScheduleCorrectionStep(
        recommended_wake_hour=tonight_wake,
        recommended_onset_hour=tonight_onset,
        shift_minutes_tonight=shift_hours * 60,
        remaining_gap_minutes=remaining,
        nights_remaining_estimate=nights_left,
        reached_target=reached,
        current_wake_hour=phase.wake_hour,
        confidence=phase.confidence,
        target_wake_hour=goal.target_wake_hour,
    )


# --- Parser ----------------------------------------------------------------
# Mirrors ScheduleGoalParser (Swift L195-291): conversational
# "up at 6am starting Monday" -> ScheduleGoal.

_SCHEDULE_CUES = (
    "up at", "wake at", "wake me", "get up at", "be up at",
    "getting up at", "alarm at", "wake time", "start waking",
    "waking up at", "need to be up",
)

_WAKE_HOUR_PATTERN = re.compile(
    r"(?:up at|wake(?: me)?(?: up)? at|get up at|be up at|getting up at|waking up at|alarm at|"
    r"wake time(?: of)?|need to be up(?: at)?)\s+(\d{1,2})(?::(\d{2}))?\s*(a\.?m\.?|p\.?m\.?)?"
)

_WEEKDAY_CUES: tuple[tuple[int, str], ...] = (
    (1, "sunday"),
    (2, "monday"),
    (3, "tuesday"),
    (4, "wednesday"),
    (5, "thursday"),
    (6, "friday"),
    (7, "saturday"),
)


def is_schedule_ask(text: str) -> bool:
    return parse(text) is not None


def parse(text: str, now: datetime | None = None) -> ScheduleGoal | None:
    if now is None:
        now = datetime.now()
    lower = text.lower()
    if not _has_schedule_cue(lower):
        return None
    hour = parse_wake_hour(lower)
    if hour is None:
        return None
    cutover = _parse_cutover(lower, now)
    return ScheduleGoal(target_wake_hour=hour, cutover_start=cutover, created_at=now)


def _has_schedule_cue(lower: str) -> bool:
    return any(cue in lower for cue in _SCHEDULE_CUES)


def parse_wake_hour(lower: str) -> float | None:
    """Mirrors ``parseWakeHour(in:)`` (Swift L226-263) exactly, dead
    branches included: `hour` is clamped to [0, 23] a few lines above the
    `hour == 24`/`hour > 24` checks below, so those two (and the final
    "already 24h"/"hour == 12" no-op pair) can never actually fire --
    ported as-is rather than silently dropped, since removing unreachable
    code the Swift source itself carries would be an unrequested behavior
    change disguised as cleanup."""
    match = _WAKE_HOUR_PATTERN.search(lower)
    if not match:
        return None
    hour = int(match.group(1))
    minute = int(match.group(2)) if match.group(2) else 0
    stamp = match.group(3)
    is_pm: bool | None = None
    if stamp:
        if "p" in stamp:
            is_pm = True
        if "a" in stamp:
            is_pm = False

    hour = min(23, max(0, hour))
    minute = min(59, max(0, minute))
    if is_pm is not None:
        if is_pm and hour < 12:
            hour += 12
        if is_pm is False and hour == 12:
            hour = 0
    elif hour == 0 or hour == 24:  # unreachable: hour already clamped to <= 23
        hour = 0
    elif hour > 24:  # unreachable: hour already clamped to <= 23
        return None
    # Bare "6" on a wake ask is 6am, not 18:00. Both arms below are no-ops
    # given the clamp above -- ported for fidelity, not because they do
    # anything observable.
    if is_pm is None and 13 <= hour <= 23:
        pass
    elif is_pm is None and hour == 12:
        hour = 12

    return float(hour) + minute / 60.0


def _parse_cutover(lower: str, now: datetime) -> datetime:
    for code, name in _WEEKDAY_CUES:
        if name in lower:
            return _next_weekday(code, now)
    return now


def _next_weekday(weekday_code: int, from_: datetime) -> datetime:
    """`weekday_code` follows Foundation's ``Calendar.component(.weekday:)``
    numbering (Sunday=1 ... Saturday=7), matching the Swift source's own
    table exactly; converted here to Python's ``datetime.weekday()``
    numbering (Monday=0 ... Sunday=6) via `(code + 5) % 7`."""
    start = from_.replace(hour=0, minute=0, second=0, microsecond=0)
    target = (weekday_code + 5) % 7
    for offset in range(8):
        day = start + timedelta(days=offset)
        if day.weekday() == target:
            return day
    return start
